### PR TITLE
Add at::tensor(complex) and torch::tensor(complex) overload

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: test/cpp/api/tensor.cpp [hunk 1]
```diff
@@ -166,6 +166,16 @@ TEST(TensorTest, AtTensorCtorScalar) {
   ASSERT_EQ(tensor.numel(), 1);
   ASSERT_EQ(tensor.dtype(), at::kFloat);
   ASSERT_TRUE(almost_equal(tensor[0], 123.5));
+
+  tensor = at::tensor(c10::complex<float>(1.0, 2.0), at::dtype(at::kComplexFloat)) + 0.5;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply test/cpp/api/tensor.cpp:170]: Good point, both should be supported and tested

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply test/cpp/api/tensor.cpp:175]: Fixed

**Comment by zasdfgbnm** [comment]
@anjali411 The test is passing on my local box, hopefully this is working.
Also cc: @yf225 for C++ API change

**Comment by yf225** [comment]
cc. @glaringlee for C++ API change :D

File: test/cpp/api/tensor.cpp [hunk 2]
```diff
@@ -17,7 +17,7 @@ bool exactly_equal(at::Tensor left, T right) {
 }
 
 template <typename T>
-bool almost_equal(at::Tensor left, T right, T tolerance = 1e-4) {
+bool almost_equal(at::Tensor left, T right, double tolerance = 1e-4) {
```
Comment (anjali411) [inline test/cpp/api/tensor.cpp:20]: why is this changed to `double`?

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -967,6 +967,22 @@ Tensor tensor_backend(ArrayRef<T> values, const TensorOptions& options) {
   return cpu_tensor.to(options.device());
 }
 
+template <typename T>
+Tensor tensor_complex_cpu(ArrayRef<T> values, const TensorOptions& options) {
```
Comment (anjali411) [inline aten/src/ATen/native/TensorFactories.cpp:971]: why do we need this? `tensor_cpu` above is already being dispatched for complex and this function looks to be doing the exact thing as above

File: test/cpp/api/tensor.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply test/cpp/api/tensor.cpp:20]: Because we can not do
```C++
double < complex
```

File: aten/src/ATen/native/TensorFactories.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/TensorFactories.cpp:971]: This is for converting **from complex to complex**.
`tensor_cpu` supports converting **from real to any dtype**.

We can not convert **from complex to any dtype**, it will fail with compile error like "can not convert c10::complex<float> to int".

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h [hunk 4]
```diff
@@ -225,6 +228,11 @@ AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TENSOR)
       fill_tensor(tensor);
       return tensor.to(options.device());
     } else if (is_tensor()) {
+      bool from_complex = tensor_.scalar_type() == at::kComplexFloat || tensor_.scalar_type() == at::kComplexDouble;
```
Comment (anjali411) [inline torch/csrc/api/include/torch/detail/TensorDataContainer.h:231]: you could use `is_complex` here:
```
bool from_complex = tensor_.is_complex()
```

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h [hunk 5]
```diff
@@ -225,6 +228,11 @@ AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TENSOR)
...
+      bool to_complex = options.dtype() == at::kComplexFloat || options.dtype() == at::kComplexDouble;
```
Comment (anjali411) [inline torch/csrc/api/include/torch/detail/TensorDataContainer.h:232]: same as above

File: test/cpp/api/tensor.cpp (hunk 2 above)
Comment (anjali411) [reply test/cpp/api/tensor.cpp:20]: ahh I see for `c10::complex<U>` the tolerance would be `complex` otherwise

File: aten/src/ATen/native/TensorFactories.cpp (hunk 3 above)
Comment (anjali411) [reply aten/src/ATen/native/TensorFactories.cpp:971]: isn't that because in the `TENSOR` macro we don't have complex dtypes included?
```
AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
```

Comment (zasdfgbnm) [reply aten/src/ATen/native/TensorFactories.cpp:971]: I don't think so, I believe it is in:
```C++
std::copy(values.begin(), values.end(), result.template data_ptr<scalar_t>());
```
where `values` is complex type, and `result.template data_ptr<scalar_t>()` is int pointer. And internally C++ does
```C++
*ptr = value;
```
but this is wrong because we can not convert complex to int.

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h]: fixed

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h:232]: Does tensor options have `is_complex`?

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h [hunk 6]
```diff
@@ -225,6 +228,11 @@ AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TENSOR)
       fill_tensor(tensor);
       return tensor.to(options.device());
     } else if (is_tensor()) {
+      bool from_complex = tensor_.is_complex();
+      bool to_complex = options.dtype() == at::kComplexFloat || options.dtype() == at::kComplexDouble;
+      if (from_complex && !to_complex) {
+        throw std::runtime_error("can not do torch::tensor(complex, dtype=non-complex) because complex can not be casted to real number without loss of information");
```
Comment (glaringlee) [inline torch/csrc/api/include/torch/detail/TensorDataContainer.h:234]: Is there any reason that we throw std::runtime_error directly here instead of using TORCH_CHECK(false.... or TORCH_INTERNAL_ASSERT(false.....?

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 5 above)
Comment (anjali411) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h:232]: no but you can do `bool to_complex = at::empty({0}, options).is_complex()`

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 6 above)
Comment (zasdfgbnm) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h:234]: You are right, I should change to `TORCH_CHECK`

Comment (zasdfgbnm) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h]: fixed

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h]: fixed

**Comment by zasdfgbnm** [comment]
This should be ready

**Review by anjali411 — APPROVED** [review]
